In [199]:
import sys
!{sys.executable} -m pip install numpy
!{sys.executable} -m pip install matplotlib
!{sys.executable} -m pip install pandas
!{sys.executable} -m pip install seaborn

# import sys
# !conda install --yes --prefix {sys.prefix} numpy
# !conda install --yes --prefix {sys.prefix} matplotlib
# !conda install --yes --prefix {sys.prefix} pandas

In [200]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

# THE DATASET
This is the dataset downloaded from the WHO website. It shows how the blood pressure in countries has varied from 1975 to 2015. Each year has seperate columns for 'Both sexes', 'Male' & 'Female'. The values in the table are in the string format: '{average} [{range}]'. The data will need to be cleaned before it can be visualized

Objectives:
1. Rename the columns so they display the information from the 3rd row.
2. Remove the first 3 rows.
3. Make the country column the index column.
4. Convert blood pressure data from having average and range to just having average
5. Convert the blood pressure data from strings to floats
6. Create a smaller dataframe only containing the (Both sexes) data
7. Transpose the data so it fits seaborn's lineplotmethod.

In [201]:
df = pd.read_csv('data.csv')
df.head()

Unnamed: 0                                               2015  \
0          NaN   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...   
1          NaN                                         18+  years   
2      Country                                         Both sexes   
3  Afghanistan                                   30.6 [23.6-38.3]   
4      Albania                                   29.0 [22.4-36.2]   

                                              2015.1  \
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...   
1                                         18+  years   
2                                               Male   
3                                   30.4 [20.4-41.6]   
4                                   33.0 [23.0-44.3]   

                                              2015.2  \
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...   
1                                         18+  years   
2                                             Female   
3                                   30.7 [21.2-41.3]   
4                                   25.0 [16.8-34.7]   

                                                2014  ...  \
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...  ...   
1                                         18+  years  ...   
2                                         Both sexes  ...   
3                                   30.6 [23.9-37.8]  ...   
4                                   29.2 [23.0-36.0]  ...   

                                              1976.1  \
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...   
1                                         18+  years   
2                                               Male   
3                                   25.2 [14.5-38.5]   
4                                   34.5 [22.4-47.7]   

                                              1976.2  \
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...   
1                                         18+  years   
2                                             Female   
3                                   23.4 [13.3-35.6]   
4                                   31.1 [20.1-44.2]   

                                                1975  \
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...   
1                                         18+  years   
2                                         Both sexes   
3                                   24.2 [16.4-33.3]   
4                                   33.1 [24.2-42.6]   

                                              1975.1  \
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...   
1                                         18+  years   
2                                               Male   
3                                   25.1 [13.9-38.9]   
4                                   34.6 [21.9-48.6]   

                                              1975.2  
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...  
1                                         18+  years  
2                                             Female  
3                                   23.2 [12.7-35.9]  
4                                   31.3 [19.7-45.1]  

[5 rows x 124 columns]

# 1. Renaming columns
We will create a function which takes the old column name as input and returns a column name with the year and information from row 3.

NOTE: we want the 'Unnamed: 0' column to be the 'Country' column

In [202]:
def get_column_name(column):
    if column.name == 'Unnamed: 0':
        return 'Country'
    else:
        return column.name[:4] + f' ({column.iloc[2][1:]})'

Now let's assign the columns property of the dataframe (df.columns) to a list. This list will be made using list comprehension (see Section 2). Each element in the list is formed by applying the get_column_name function on the current element in df.columns.

In [203]:
df.columns = [get_column_name(df[column_name]) for column_name in df.columns]

# Removing the 1st 3 rows

In [204]:
df.iloc[:3]

Country                                  2015 (Both sexes)  \
0      NaN   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...   
1      NaN                                         18+  years   
2  Country                                         Both sexes   

                                         2015 (Male)  \
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...   
1                                         18+  years   
2                                               Male   

                                       2015 (Female)  \
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...   
1                                         18+  years   
2                                             Female   

                                   2014 (Both sexes)  ...  \
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...  ...   
1                                         18+  years  ...   
2                                         Both sexes  ...   

                                         1976 (Male)  \
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...   
1                                         18+  years   
2                                               Male   

                                       1976 (Female)  \
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...   
1                                         18+  years   
2                                             Female   

                                   1975 (Both sexes)  \
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...   
1                                         18+  years   
2                                         Both sexes   

                                         1975 (Male)  \
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...   
1                                         18+  years   
2                                               Male   

                                       1975 (Female)  
0   Raised blood pressure (SBP&gt;=140 OR DBP&gt;...  
1                                         18+  years  
2                                             Female  

[3 rows x 124 columns]

As you can see from above, the 3rd row of the DataFrame is useless now that we have renamed the columns. The 1st and 2nd rows were always useless. Let's remove them:

In [205]:
df = df.iloc[3:]
df.head()

Country 2015 (Both sexes)       2015 (Male)     2015 (Female)  \
3  Afghanistan  30.6 [23.6-38.3]  30.4 [20.4-41.6]  30.7 [21.2-41.3]   
4      Albania  29.0 [22.4-36.2]  33.0 [23.0-44.3]  25.0 [16.8-34.7]   
5      Algeria  25.1 [19.4-31.5]  25.4 [17.1-35.2]  24.6 [16.6-33.6]   
6      Andorra  18.7 [13.3-24.8]  23.2 [14.3-33.5]   14.2 [8.3-21.5]   
7       Angola  29.7 [22.1-38.3]  29.6 [18.1-42.7]  29.6 [19.2-41.3]   

  2014 (Both sexes)  ...       1976 (Male)     1976 (Female)  \
3  30.6 [23.9-37.8]  ...  25.2 [14.5-38.5]  23.4 [13.3-35.6]   
4  29.2 [23.0-36.0]  ...  34.5 [22.4-47.7]  31.1 [20.1-44.2]   
5  25.4 [20.0-31.5]  ...  30.6 [18.2-44.1]  32.1 [20.4-45.5]   
6  19.2 [14.0-24.9]  ...  41.3 [28.8-53.6]  34.4 [23.3-46.5]   
7  29.8 [22.6-38.0]  ...  28.5 [15.8-44.1]  25.9 [14.8-39.4]   

  1975 (Both sexes)       1975 (Male)     1975 (Female)  
3  24.2 [16.4-33.3]  25.1 [13.9-38.9]  23.2 [12.7-35.9]  
4  33.1 [24.2-42.6]  34.6 [21.9-48.6]  31.3 [19.7-45.1]  
5  31.5 [22.5-41.2]  30.7 [17.6-45.0]  32.1 [19.6-46.1]  
6  37.9 [29.1-47.1]  41.3 [28.1-54.4]  34.6 [22.9-47.3]  
7  27.0 [18.2-37.3]  28.3 [15.2-44.6]  25.6 [14.2-39.9]  

[5 rows x 124 columns]

# Making the 'Country' column the index column of the DataFrame
This is so that we can now access rows from their corresponding country using df.loc[country_name]

In [206]:
df = df.set_index('Country')

df.head()

2015 (Both sexes)       2015 (Male)     2015 (Female)  \
Country                                                             
Afghanistan  30.6 [23.6-38.3]  30.4 [20.4-41.6]  30.7 [21.2-41.3]   
Albania      29.0 [22.4-36.2]  33.0 [23.0-44.3]  25.0 [16.8-34.7]   
Algeria      25.1 [19.4-31.5]  25.4 [17.1-35.2]  24.6 [16.6-33.6]   
Andorra      18.7 [13.3-24.8]  23.2 [14.3-33.5]   14.2 [8.3-21.5]   
Angola       29.7 [22.1-38.3]  29.6 [18.1-42.7]  29.6 [19.2-41.3]   

            2014 (Both sexes)       2014 (Male)  ...       1976 (Male)  \
Country                                          ...                     
Afghanistan  30.6 [23.9-37.8]  30.4 [21.0-41.0]  ...  25.2 [14.5-38.5]   
Albania      29.2 [23.0-36.0]  33.2 [23.9-43.7]  ...  34.5 [22.4-47.7]   
Algeria      25.4 [20.0-31.5]  25.7 [17.9-34.9]  ...  30.6 [18.2-44.1]   
Andorra      19.2 [14.0-24.9]  23.7 [15.1-33.4]  ...  41.3 [28.8-53.6]   
Angola       29.8 [22.6-38.0]  29.8 [18.8-42.2]  ...  28.5 [15.8-44.1]   

                1976 (Female) 1975 (Both sexes)       1975 (Male)  \
Country                                                             
Afghanistan  23.4 [13.3-35.6]  24.2 [16.4-33.3]  25.1 [13.9-38.9]   
Albania      31.1 [20.1-44.2]  33.1 [24.2-42.6]  34.6 [21.9-48.6]   
Algeria      32.1 [20.4-45.5]  31.5 [22.5-41.2]  30.7 [17.6-45.0]   
Andorra      34.4 [23.3-46.5]  37.9 [29.1-47.1]  41.3 [28.1-54.4]   
Angola       25.9 [14.8-39.4]  27.0 [18.2-37.3]  28.3 [15.2-44.6]   

                1975 (Female)  
Country                        
Afghanistan  23.2 [12.7-35.9]  
Albania      31.3 [19.7-45.1]  
Algeria      32.1 [19.6-46.1]  
Andorra      34.6 [22.9-47.3]  
Angola       25.6 [14.2-39.9]  

[5 rows x 123 columns]

### Challenge: Access 'Andorra' row using .loc[]

In [207]:
df.loc['Albania']

2015 (Both sexes)    29.0 [22.4-36.2]
2015 (Male)          33.0 [23.0-44.3]
2015 (Female)        25.0 [16.8-34.7]
2014 (Both sexes)    29.2 [23.0-36.0]
2014 (Male)          33.2 [23.9-43.7]
                           ...       
1976 (Male)          34.5 [22.4-47.7]
1976 (Female)        31.1 [20.1-44.2]
1975 (Both sexes)    33.1 [24.2-42.6]
1975 (Male)          34.6 [21.9-48.6]
1975 (Female)        31.3 [19.7-45.1]
Name: Albania, Length: 123, dtype: object

# Removing the range part of the blood pressure data

1. Below we define a function which takes in a cell from the DataFrame as input. 
2. This cell will be a string, e.g.'18.6 [13.7-24.1]'. cell.split(' ') returns a list of strings formed by splitting up the cell string at the spaces giving: ['18.6', '[13.7-24.1]'].
3. We then want the first element (0th index) of this list (as this is the mean blood pressure) so we use: cell.split(' ')[0]

In [208]:
def remove_range(cell):
    return cell.split(' ')[0]

## Using .apply() to change the values in a DataFrame
Below we use a for loop to pass through every column name in df.columns. For each column name, we set the column's DataFrame (df[column_name]) equal to df[column_name].apply(remove_range). This is like setting it equal to the original column's DataFrame after we have 'applied' the remove_range(cell) function on every cell in the column.

In [209]:
for column_name in df.columns:
    df[column_name] = df[column_name].apply(remove_range)
df

2015 (Both sexes) 2015 (Male)  \
Country                                                            
Afghanistan                                     30.6        30.4   
Albania                                         29.0        33.0   
Algeria                                         25.1        25.4   
Andorra                                         18.7        23.2   
Angola                                          29.7        29.6   
...                                              ...         ...   
Venezuela (Bolivarian Republic of)              18.6        21.5   
Viet Nam                                        23.4        25.0   
Yemen                                           30.7        29.9   
Zambia                                          27.1        27.6   
Zimbabwe                                        28.2        26.9   

                                   2015 (Female) 2014 (Both sexes)  \
Country                                                              
Afghanistan                                 30.7              30.6   
Albania                                     25.0              29.2   
Algeria                                     24.6              25.4   
Andorra                                     14.2              19.2   
Angola                                      29.6              29.8   
...                                          ...               ...   
Venezuela (Bolivarian Republic of)          15.7              19.0   
Viet Nam                                    21.6              23.4   
Yemen                                       31.2              30.7   
Zambia                                      26.5              27.2   
Zimbabwe                                    29.2              28.3   

                                   2014 (Male)  ... 1976 (Male) 1976 (Female)  \
Country                                         ...                             
Afghanistan                               30.4  ...        25.2          23.4   
Albania                                   33.2  ...        34.5          31.1   
Algeria                                   25.7  ...        30.6          32.1   
Andorra                                   23.7  ...        41.3          34.4   
Angola                                    29.8  ...        28.5          25.9   
...                                        ...  ...         ...           ...   
Venezuela (Bolivarian Republic of)        21.9  ...        35.1          33.5   
Viet Nam                                  24.9  ...        21.7          18.8   
Yemen                                     29.8  ...        25.4          26.3   
Zambia                                    27.7  ...        30.5          27.2   
Zimbabwe                                  27.0  ...        28.5          27.4   

                                   1975 (Both sexes) 1975 (Male) 1975 (Female)  
Country                                                                         
Afghanistan                                     24.2        25.1          23.2  
Albania                                         33.1        34.6          31.3  
Algeria                                         31.5        30.7          32.1  
Andorra                                         37.9        41.3          34.6  
Angola                                          27.0        28.3          25.6  
...                                              ...         ...           ...  
Venezuela (Bolivarian Republic of)              34.5        35.1          33.6  
Viet Nam                                        20.3        21.7          18.7  
Yemen                                           25.8        25.4          26.1  
Zambia                                          28.7        30.3          27.0  
Zimbabwe                                        27.8        28.3          27.2  

[195 rows x 123 columns]

## Trying (and failing) to convert the data from strings to floats

We use the pd.to_numeric method to convertn cells in df_both columns from strings into floats

In [210]:
for column_name in df.columns:
    df[column_name] = pd.to_numeric(df[column_name], downcast='float')

ValueError: Unable to parse string "No" at position 112

However, we get an error...

# Realising that for some countries, there is "No" data:

Some of the rows do not contain any data, only the string 'No'. For example Sudan:

In [211]:
df.loc['Sudan']

2015 (Both sexes)    No
2015 (Male)          No
2015 (Female)        No
2014 (Both sexes)    No
2014 (Male)          No
                     ..
1976 (Male)          No
1976 (Female)        No
1975 (Both sexes)    No
1975 (Male)          No
1975 (Female)        No
Name: Sudan, Length: 123, dtype: object

This isn't very helpful for plotting data so these rows should be removed

# Finding these countries
First, lets create a list of all countries. We can do this using df_both.index since the Country column is the DataFrame's index column

In [ ]:
countries = df.index

First, create an empty list called countries_to_remove. Now use for loop to pass through each country in countries list above. And if the value at the row of that country in the '2015' column equals 'No', then we add it to countres_to_remove

In [ ]:
countries_to_remove = []

for country in countries:
    if df_both.loc[country]['2015'] == 'No':
        countries_to_remove.append(country)
        
countries_to_remove

# Removing these countries
We now drop those countries.We again use .drop() but as labels is now set equal to names of rows, axis = 0.

In [ ]:
df = df.drop(labels=countries_to_remove, axis=0)

# Converting the string values to floats
Since we have removed the 'No' strings, we can convert the number strings, e.g. '30.6' into floats. We pass through each column_name df.columns with a for loop. For each, we assign the DataFrame at that column name to itself after we have applied the pd.to_numeric function with downcast='float'

In [ ]:
for column_name in df.columns:
    df[column_name] = pd.to_numeric(df[column_name], downcast='float')

## Creating a new dataset (df_both) which only contains the (Both sexes) Columns from df

Below we will create an empty list called column_names_of_columns_to_drop. Then, we use a for loop to iterate through the column names in df.columns. If the 2nd element (1st index) of the column_name.split(' ') is not equal to '(Both', we know it is not a 'Both sexes' column, so we can add that to the list of column names of columns we want to drop from the DataFrame.

In [ ]:
column_names_of_columns_to_drop = []

for column_name in df.columns:
    if column_name.split(' ')[1] != '(Both':
        column_names_of_columns_to_drop.append(column_name)

We can now create our new DataFrame, df_both, which we will set equal to the original DataFrame df with .drop() applied on it. The labels (names) of columns to drop is set equal to our new list and axis = 1.

In [ ]:
df_both = df.drop(labels=column_names_of_columns_to_drop, axis=1)

### Challenge: Rename the columns in df_both
We can now rename the columns because they are 'Both sexes' columns.

In [ ]:
df_both.columns = [column_name.split(' ')[0] for column_name in df_both.columns]

## Transposing the data
In order to use the lineplot function from seaborn, the years must be the rows and countries the columns. This is so that the years appear on the x-axis of the plot.

In [232]:
df_both = df_both.transpose()

### Reversing the rows
NOTE: As the years in our dataframe start at 2015 and descend down to 1975, this is how it gets shown on the lineplot, when in fact we want the years to be ascending from left to right. To fix this, we can reverse the rows of our dataframe using list slicing: df_both.iloc[::-1]

In [234]:
df_both = df_both.iloc[::-1]

df_both

Country  Afghanistan    Albania    Algeria    Andorra     Angola  ...  \
1975       24.200001  33.099998  31.500000  37.900002  27.000000  ...   
1976       24.400000  33.000000  31.500000  37.900002  27.200001  ...   
1977       24.600000  32.799999  31.500000  37.700001  27.500000  ...   
1978       24.799999  32.700001  31.500000  37.599998  27.700001  ...   
1979       25.100000  32.599998  31.500000  37.400002  27.900000  ...   
...              ...        ...        ...        ...        ...  ...   
2011       30.400000  29.900000  26.500000  20.600000  30.200001  ...   
2012       30.500000  29.700001  26.200001  20.100000  30.100000  ...   
2013       30.500000  29.500000  25.799999  19.600000  30.000000  ...   
2014       30.600000  29.200001  25.400000  19.200001  29.799999  ...   
2015       30.600000  29.000000  25.100000  18.700001  29.700001  ...   

Country  Venezuela (Bolivarian Republic of)   Viet Nam      Yemen     Zambia  \
1975                              34.500000  20.299999  25.799999  28.700001   
1976                              34.400002  20.299999  25.900000  28.900000   
1977                              34.200001  20.400000  26.100000  29.000000   
1978                              34.099998  20.400000  26.200001  29.200001   
1979                              33.900002  20.400000  26.400000  29.299999   
...                                     ...        ...        ...        ...   
2011                              20.200001  23.200001  30.600000  27.400000   
2012                              19.799999  23.200001  30.600000  27.299999   
2013                              19.400000  23.299999  30.700001  27.299999   
2014                              19.000000  23.400000  30.700001  27.200001   
2015                              18.600000  23.400000  30.700001  27.100000   

Country   Zimbabwe  
1975     27.799999  
1976     28.000000  
1977     28.200001  
1978     28.400000  
1979     28.600000  
...            ...  
2011     28.600000  
2012     28.500000  
2013     28.400000  
2014     28.299999  
2015     28.200001  

[41 rows x 191 columns]

## Creating a line plotting function which takes countries to plot as arguments

### Setting the figsize, title and axes labels using matplotlib (plt)
We use matplotlib to set the dimensions of our graph and to give it a title & x-axis and y-axis labels. 

In [236]:
def dimensions_and_naming():
    plt.figure(figsize=(10,6))

    plt.title("How has blood pressure has varied in different countries between 1975 and 2015")
    plt.ylabel('Blood pressure')
    plt.xlabel('Year')

### Setting the year xticks in increments of 5
As we have data for every year from over 40 years, the graph will try to fit every year on the x axis, leaving it overcrowded. To solve this problem, we can use the numpy external library (np).

We will use the np.arange() method to create a list of numbers. We need to create a list of numbers which starts at 0 and ends at 40 and has increments of 5. np.arange() takes similar arguments to those of list slicing. We want the start to be 0, the stop to be 41 and the step to be 5. So we write np.arange(0, 41, 5).

This is because between 1975 and 2015 there are 40 years and on the x axis we only want to show 1 year in every 5.

In [ ]:
def setting_xticks():
    lst = np.arange(0,41,5)
    plt.xticks(lst)

### Using seaborn (sns) lineplot
To plot a line graph, an option is to use the seaborn lineplot method. The keyword argument is data which you give the dataframe you would like to plot. This data must be correctly formatted with the variable you wish to be on the x-axis being in the rows of the dataframe.

NOTE: When we plot all the data for every country, the graph is very messy. To select only a few countries, we can select them by their column name. To select multiple columns from their name, we use dataframe double square brackets and within the inner square bracket, we write the list of country names.

In [235]:
dimensions_and_naming()

setting_xticks()

sns.lineplot(data=df_both[['Uruguay', 'United States of America', 'Zambia', 'Zimbabwe']])

NameError: name 'dimensions_and_naming' is not defined